Tutorial revisions (#415)
JuliaGeo/GeometryOps.jl: docs/src/tutorials/geospatial_geometry.md

import GeoFormatTypes as GFT
using GeoParquet
using GeoDataFrames
# Packages for coordinate transformation and projection
import CoordinateTransformations
import Proj
# Plotting
using CairoMakie
using GeoMakie
using DisplayAs # hide
Makie.set_theme!(Makie.MAKIE_DEFAULT_THEME) # hide
# Loading data
using GeoJSON 
using DataFrames
import Shapefile
````

## [Plot geometries on a map using `GeoMakie` and coordinate reference system (`CRS`)](@id plot-geometry) 

In geospatial sciences we often have data in one [Coordinate Reference System (CRS)](https://en.wikipedia.org/wiki/Spatial_reference_system) (`source`) and would like to display it in different (`destination`) `CRS`. `GeoMakie` allows us to do this by automatically projecting from `source` to `destination` CRS.

Here, our `source` CRS is common geographic (i.e. coordinates of latitude and longitude), [WGS84](https://epsg.io/4326).

````@example geospatial_geometry
source_crs1 = GFT.EPSG(4326)
````

Now let's pick a `destination` CRS for displaying our map. Here we'll pick [natearth2](https://proj.org/en/9.4/operations/projections/natearth2.html).

````@example geospatial_geometry
destination_crs = "+proj=natearth2"
````

> Let's add land area for context. First, download and open the [Natural Earth](https://www.naturalearthdata.com) global land polygons at 110 m resolution.`GeoMakie` ships with this particular dataset, so we will access it from there.

Reviewer: ```suggestion
Let's add land area for context. First, download and open the [Natural Earth](https://www.naturalearthdata.com) global land polygons at 110 m resolution.  `GeoMakie` ships with this particular dataset, so we will access it from there.
Note that this will be a path on your local machine, so you could easily point it to any other `.geojson` file you have.
```

````@example geospatial_geometry
land_path = GeoMakie.assetpath("ne_110m_land.geojson")
````

!!! note
    Natural Earth has lots of other datasets, and there is a Julia package that provides an interface to it called [NaturalEarth.jl](https://github.com/JuliaGeo/NaturalEarth.jl).

> Read the land `MultiPolygon`s as a `GeoJSON.FeatureCollection`.

Reviewer: ```suggestion
Read this dataset in as a `GeoJSON.FeatureCollection`.
```

````@example geospatial_geometry
land_geo = GeoJSON.read(land_path)
````

> We then need to create a figure with a `GeoAxis` that can handle the projection between `source` and `destination` CRS. For GeoMakie, `source` is the CRS of the input and `dest` is the CRS you want to visualize in.

Reviewer: Could be good to cross link here to the GeoMakie docs

````@example geospatial_geometry
fig = Figure(size=(1000, 500));
ga = GeoAxis(
    fig[1, 1];
    source = source_crs1,
    dest = destination_crs,
    xticklabelsvisible = false,
    yticklabelsvisible = false,
);
nothing #hide
````

Plot `land` for context.

````@example geospatial_geometry

> poly!(ga, land_geo, color=:black)

Reviewer: ```suggestion
poly!(ga, land_geo; color=:black)
```

fig
````

Now let's plot a `Polygon` like before, but this time with a CRS that differs from our `source` data

````@example geospatial_geometry
ϴ = 0:0.01:2π

> ring(r) = GI.LinearRing(GI.Point.(zip(50 .+ r .* cos.(ϴ), 50 .+ r .* sin.(ϴ))))

Reviewer: ```suggestion
function ring(radius)
    ϴ = 0:0.01:2π
	points = @. GI.Point(50 + r * cos(ϴ), 50 + r * sin(ϴ))
	return GI.LinearRing(points)
end
```

> spiro = GI.LinearRing(GI.Point.(zip(50 .+ 22cos.(ϴ) .- 2cos.(11ϴ), 50 .+ 22sin.(ϴ) .- 2sin.(11ϴ))))

Reviewer: similar simplification for this please, otherwise it's a bit hard to parse

Reviewer: ("this" => spiro)

fig
````

## [Create geospatial geometries with embedded coordinate reference system information](@id geom-crs)
 
Great, we can make geometries and plot them on a map... now let's export the data to common geospatial data formats. To do this we now need to create geometries with embedded `CRS` information, making it a geospatial geometry. All that's needed is to include `; crs = crs` as a keyword argument when constructing the geometry.

Let's do this for a new `Polygon`
````@example geospatial_geometry
r = 3;
k = 7;
ϴ = 0:0.01:2pi;
x = r .* (k + 1) .* cos.(ϴ) .- r .* cos.((k + 1) .* ϴ);
y = r .* (k + 1) .* sin.(ϴ) .- r .* sin.((k + 1) .* ϴ);
ring4 = GI.LinearRing(Point.(zip(x, y)))
````

But this time when we create the `Polygon` we need to specify the `CRS` at the time of creation, making it a geospatial polygon

````@example geospatial_geometry
geopoly1 = GI.Polygon([ring4], crs = source_crs1)
````

!!! note
    It is good practice to only include CRS information with the highest-level geometry. Not doing so can bloat the memory footprint of the geometry. CRS information _can_ be included at the individual `Point` level but is discouraged.

> And let's create second `Polygon` by shifting the first using CoordinateTransformations

Reviewer: ```suggestion
And let's create a second `Polygon` by shifting the first using CoordinateTransformations
```

````@example geospatial_geometry
xoffset = 20.;
yoffset = -25.;
f = CoordinateTransformations.Translation(xoffset, yoffset);

> geopoly2 = GO.transform(f, geopoly1);

Reviewer: ```suggestion
xoffset = 20.
yoffset = -25.
f = CoordinateTransformations.Translation(xoffset, yoffset)
geopoly2 = GO.transform(f, geopoly1)
```

why the semicolons?

````

## [Creating a table with attributes and geometry](@id attributes)

Typically, you'll also want to include attributes with your geometries. Attributes are simply data that are attributed to each geometry. The easiest way to do this is to create a table with a `:geometry` column. Let's do this using [`DataFrames`](https://github.com/JuliaData/DataFrames.jl).

````@example geospatial_geometry
df = DataFrame(geometry=[geopoly1, geopoly2])
````

> Now let's add a couple of attributes to the geometries.  We do this using [DataFrames' `!` mutation syntax](https://dataframes.juliadata.org/stable/man/getting_started/#The-DataFrame-Type) that allows you to add a new column to an existing data frame.

Reviewer: nice, I like the cross link there

Reviewer: fyi, this can also be `df.id = ...; df.name = ...`

We begin with [GeoJSON](https://github.com/JuliaGeo/GeoJSON.jl), which is a [JSON](https://en.wikipedia.org/wiki/JSON) format for geospatial feature collections.  It's human-readable and widely supported by most web-based and desktop geospatial libraries.

````@example geospatial_geometry
fn = "shapes.json"
GeoJSON.write(fn, df)
````

Now, let's save as a [`Shapefile`](https://github.com/JuliaGeo/Shapefile.jl).  Shapefiles are actually a set of files (usually 4) that hold geometry information, a CRS, and additional attribute information as a separate table.  When you give `Shapefile.write` a file name, it will write 4 files of the same name but with different extensions.

````@example geospatial_geometry
fn = "shapes.shp"
Shapefile.write(fn, df)
````

Now, let's save as a [`GeoParquet`](https://github.com/JuliaGeo/GeoParquet.jl).  GeoParquet is a geospatial extension to the [Parquet](https://parquet.apache.org/) format, which is a high-performance data store.  It's great for storing large amounts of data in a single file.

````@example geospatial_geometry
fn = "shapes.parquet"

> GeoParquet.write(fn, df, (:geometry,))

Reviewer: ```suggestion
GeoParquet.write(fn, df)
```

I rationalized this syntax in https://github.com/JuliaGeo/GeoParquet.jl/commit/2e52f16b1bacb4d7de98465585cbf71672158c07

also, do we want to keep doing this `fn = ...; YourPkg.write(fn, df)`?  Or just do `YourPkg.write("shapes.parquet", df)`. 

It's also worth considering whether we should put all of this into a tabset or collapsible block, and recommend GeoDataFrames as the primary way to do this, to avoid complexity.

````

> Finally, if there's no Julia-native package that can write data to your desired format (e.g. `.gpkg`, `.gml`, etc), you can use [`GeoDataFrames`](https://github.com/evetion/GeoDataFrames.jl). This package uses the [GDAL](https://gdal.org/) library under the hood which supports writing to nearly all geospatial formats.

Reviewer: I might have written this?  It's now outdated though since GeoDataFrames integrates with all the native IO packages if they're loaded.  So to write GeoJSON, just do

```julia
import GeoDataFrames, GeoJSON

GeoDataFrames.write("file.geojson", df)
```

````@example geospatial_geometry
fn = "shapes.gpkg"
GeoDataFrames.write(fn, df)
````

## [Geodesic paths](@id geodesic-paths)

Geodesic paths are paths computed on an ellipsoid, as opposed to a plane. The geodesic is the shortest path between two points measured along the Earth's curved surface. Because the surface is curved, that shortest path appears as a curve (not a straight line) when drawn on a flat map. 

Here, we use the `segmentize` function to add vertices along the geodesic between two points, so the line follows Earth's curved surface instead of a straight line. 

````@example geospatial_geometry
# Two points in (longitude, latitude) order: Houston (IAH) and Amsterdam (AMS).
# Geodesic methods assume lon/lat input.
IAH = (-95.358421, 29.749907)
AMS = (4.897070, 52.377956)

# Draw coastlines for geographic context
fig, ga, _cp = lines(GeoMakie.coastlines(); axis = (; type = GeoAxis))

# Create our line along the Earth, accounting for curvature

> lines!(ga, GO.segmentize(GO.Geodesic(), GI.LineString([IAH, AMS]); max_distance = 100_000); color = Makie.wong_colors()[2])

Reviewer: ```suggestion
lines!(ga, GO.segmentize(GO.Geodesic(), GI.LineString([IAH, AMS]); max_distance = 100_000))
```

probably this didn't work a while ago and now it does
